src/compiler/checker.ts

```diff
@@ -7157,6 +7183,17 @@ namespace ts {
                 propertiesArray.push(member);
             }
 
+            // If object literal is contextually typed by the implied type of a binding pattern, augment the result
+            // type with those properties for which the binding pattern specifies a default value.
+            if (contextualType && contextualType.flags & TypeFlags.ImpliedType) {

```

Comment:
It did error before (complaining that the object literal is missing property `a`), but it shouldn't have.